### ISSUE TITLE
output the style features

### ISSUE THREAD

**Issue by snownus** [jcjohnson/neural-style#29]
Can the code output style features and how it can be done?

**Comment by jcjohnson** [comment]
I'm not exactly sure what you mean; do you want the CxC Gram matrix of features from the style image? If so those are stored in the `self.target` field of the `nn.StyleLoss` modules:

https://github.com/jcjohnson/neural-style/blob/master/neural_style.lua#L304

All of the `nn.StyleLoss` modules are held in the `style_losses` instance variable here:

https://github.com/jcjohnson/neural-style/blob/master/neural_style.lua#L129

I'm not sue why you'd want to output the target Gram matrices, but it should be pretty easy to modify the code to do so.

**Comment by snownus** [comment]
Thanks for your answering.
Yes. I want to extract the style features and do some applications for example, image classification based on this feature. Is it possible? I haven't touched this kind of language, a little difficult for me to revise the code.

**Comment by dribnet** [comment]
Relatedly - with the new multiple style option, it looks like this is implemented internally with a list of Gram matrices. Any chance there is an operation for collapsing these down to one matrix? I'm imagining priming on many hundreds of images and then saving that single matrix as a saved style to be reloaded.

**Comment by jcjohnson** [comment]
Even with multiple style targets it only keeps one Gram matrix internally for each target layer, and adds to it for each style target image:

https://github.com/jcjohnson/neural-style/blob/master/neural_style.lua#L168

However it does keep a list of all (preprocessed) style images on the GPU; if you wanted to use hundreds of inputs then this would be a bit wasteful, and we should probably change it to load the style targets one at a time rather than all at once.

Multiple style images doesn't add too much overhead; each additional style target requires one additional forward pass through the network, which takes less time than a single optimization step (each of which takes a forward and backward pass; backward is slower).

Even so, if you had hundreds of style inputs it might be worth saving the resulting Gram matrix. I'm also really curious to know what you would generate from that!

**Comment by dribnet** [comment]
Do you think there could be way to compose two different style images into a single style image such that the effect of applying this composite target on the forward pass is equivalent to applying two two original style targets as is happening now? Seems unlikely, but would love to be wrong because it would be interesting to apply this to N images to construct their stylistic composite.

**Comment by jcjohnson** [comment]
I'm not 100% sure I parsed your question correctly, so let me rephrase and tell me if I got it right:

Right now when you pass multiple style images, we run each style image forward through the network to extract a Gram matrix, then average the Gram matrices for the style images to produce the target Gram matrix for the generated image. You want to know whether, given a set of style images, there is a way to combine them to produce a single image such that the Gram matrix computed from the combined image is the same as the average of the Gram matrices of the individual images?

If I've correctly understood, then the answer is actually pretty trivial: if you take multiple style images and combine them into a big image by tiling, then I'm pretty sure that the (normalized) Gram matrix of the tiled image will be the same as the average of the (normalized) Gram matrices of the original images.

**Comment by dribnet** [comment]
No worries, I think I was on the wrong track there. Your pointers helped me understand better.

So I think I have an initial working version of what I wanted [here](https://github.com/jcjohnson/neural-style/compare/master...dribnet:save_style). It's only about 10 lines or so of code and allows you to `save_style` and later `load_style`. You can even use both at once alongside a new set of style_images (if they match in size) to merge those into a new style.

Let me know if this looks interesting, and if so I can clean it up for a merge request. In particular, I think `save_style` should really save some collection of style variables that are reloaded at defaults - certainly at least style_layers and the target image size for the image scale.

Below is an example session (with some args omitted).

```bash
# generate a new image and save the starry style
th neural_style.lua \
  -style_image examples/inputs/starry_night_crop.png \
  -save_style starry_style \
  -content_image examples/inputs/golden_gate.jpg \
  -style_weight 1e11  \
  ...
```
![starry_bridge](https://cloud.githubusercontent.com/assets/945979/9952612/a5de8782-5d8c-11e5-836c-8f87249294bb.png)
```bash
# load saved starry style, add on scream style image, save that composite style and output image
# (note style weight now halved)
th neural_style.lua \
  -load_style starry_style \
  -style_image examples/inputs/the_scream_crop.png \
  -save_style starry_scream_style \
  -content_image examples/inputs/golden_gate.jpg \
  -style_weight 5e10  \
  ...
```
![starry_scream_bridge1](https://cloud.githubusercontent.com/assets/945979/9952628/e34e6628-5d8c-11e5-9025-13a1380920dc.png)
```bash
# now we can use just that save "starry scream" composite style (no style images)
th neural_style.lua \
  -load_style starry_scream_style \
  -style_image '' \
  -content_image examples/inputs/golden_gate.jpg \
  -style_weight 5e10  \
  ...
```
![starry_scream_bridge4](https://cloud.githubusercontent.com/assets/945979/9952858/81f954de-5d8f-11e5-865c-697d319cfa29.png)

Also to be clear, here's the disk representation:
```bash
$ ls -lh starry_scream_style*
-rw-r--r--  1 tom  staff    16K Sep 17 22:21 starry_scream_style_relu1_1.matrix
-rw-r--r--  1 tom  staff    64K Sep 17 22:21 starry_scream_style_relu2_1.matrix
-rw-r--r--  1 tom  staff   256K Sep 17 22:21 starry_scream_style_relu3_1.matrix
-rw-r--r--  1 tom  staff   1.0M Sep 17 22:21 starry_scream_style_relu4_1.matrix
-rw-r--r--  1 tom  staff   1.0M Sep 17 22:21 starry_scream_style_relu5_1.matrix
```

**Comment by dribnet** [comment]
Fun footnote - I confirmed that this composite style is undoable - you can trivially subtract off one of the styles to recover the other. Then since my image from my composite style came out dark, I decided to fix it by "subtracting off" seated nude to get:

![starry_scream_antinude_bridge1](https://cloud.githubusercontent.com/assets/945979/9953466/f01e2308-5d95-11e5-94b0-ef59c7d2f941.png)

Note that also seems to also remove areas of flat color, which would be expected. To be clear, the applied style there is:

![style_algebra](https://cloud.githubusercontent.com/assets/945979/9953408/67efc32e-5d95-11e5-9181-4f79d8f0f481.png)

Are you ready for style algebra? :-)

![mind-blown](https://cloud.githubusercontent.com/assets/945979/9953504/33753100-5d96-11e5-831d-3fbdd5d6cd82.gif)

**Comment by jcjohnson** [comment]
Wow - I don't think I'm ready for style algebra quite yet ;)

Even though I had never thought of using negative styles, it actually already works with master:

```
th neural_style.lua \
  -content_image examples/inputs/golden_gate.jpg \
  -style_image examples/inputs/starry_night.jpg,examples/inputs/the_scream.jpg \
  -style_blend_weights 1,1 \
  -content_weight 1e0 \
  -style_weight 1e2 \
  -tv_weight 1e-3 \
  -style_scale 0.8
```
![gate2](https://cloud.githubusercontent.com/assets/2718714/9954746/56ecd99e-5da0-11e5-9e0b-52a4b2a76ae8.png)

Then using the same style algebra equation as you:

```
th neural_style.lua \
  -content_image examples/inputs/golden_gate.jpg \
  -style_image examples/inputs/starry_night.jpg,examples/inputs/the_scream.jpg,examples/inputs/seated-nude.jpg \
  -style_blend_weights 1,1,-1 \
```

![gate3](https://cloud.githubusercontent.com/assets/2718714/9954747/56edeb9a-5da0-11e5-8096-74b4d5c41e97.png)

I guess I'm not really seeing the use case for saving style targets to disk - is there something you can do by saving and loading that you can't do by just recomputing the composite style at runtime? Saving and loading seems like more of a performance optimization unless I'm missing something.
